{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manipolazione di un file GTF (Gene Transfer Format) con `Pandas`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importare `Pandas` e leggere il file `GTF`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('./input.gtf', sep = '\\t', header = None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Cambiare i nomi delle colonne\n",
    "\n",
    "I nomi delle colonne devono essere:\n",
    "- reference\n",
    "- source\n",
    "- feature\n",
    "- start\n",
    "- end\n",
    "- score\n",
    "- strand\n",
    "- frame\n",
    "- attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "replace_dict = {0 : 'reference', 1 : 'source', 2 : 'feature', 3 : 'start',\n",
    "                4 : 'end', 5 : 'score', 6 : 'strand', 7 : 'frame', 8 : 'attributes'}\n",
    "\n",
    "df.rename(columns = replace_dict, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference      source feature  start    end score strand frame  \\\n",
       "0      ENm006  VEGA_Known    exon  71783  71788     .      -     .   \n",
       "1      ENm006  VEGA_Known     CDS  71783  71788     .      -     0   \n",
       "2      ENm006  VEGA_Known    exon  70312  70440     .      -     .   \n",
       "3      ENm006  VEGA_Known     CDS  70312  70440     .      -     0   \n",
       "4      ENm006  VEGA_Known    exon  69989  70210     .      -     .   \n",
       "..        ...         ...     ...    ...    ...   ...    ...   ...   \n",
       "405    ENm006  VEGA_Known     CDS  56303  56327     .      +     0   \n",
       "406    ENm006  VEGA_Known    exon  56689  57573     .      +     .   \n",
       "407    ENm006  VEGA_Known     CDS  56689  57573     .      +     2   \n",
       "408    ENm006  VEGA_Known    exon  57680  58323     .      +     .   \n",
       "409    ENm006  VEGA_Known     CDS  57680  57885     .      +     2   \n",
       "\n",
       "                                            attributes  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Eliminare le colonne `source` e `score`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(['source', 'score'], axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference feature  start    end strand frame  \\\n",
       "0      ENm006    exon  71783  71788      -     .   \n",
       "1      ENm006     CDS  71783  71788      -     0   \n",
       "2      ENm006    exon  70312  70440      -     .   \n",
       "3      ENm006     CDS  70312  70440      -     0   \n",
       "4      ENm006    exon  69989  70210      -     .   \n",
       "..        ...     ...    ...    ...    ...   ...   \n",
       "405    ENm006     CDS  56303  56327      +     0   \n",
       "406    ENm006    exon  56689  57573      +     .   \n",
       "407    ENm006     CDS  56689  57573      +     2   \n",
       "408    ENm006    exon  57680  58323      +     .   \n",
       "409    ENm006     CDS  57680  57885      +     2   \n",
       "\n",
       "                                            attributes  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 7 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sostituire la colonna degli attributi con le due colonne  `gene` e `transcript`\n",
    "\n",
    "La colonne `gene` e `transcript` dovranno contenere solo il nome del gene e l'ID del trascritto."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['transcript_id'] = df['attributes'].apply(lambda x : re.search(r'transcript_id\\s+\"(.+?)\";', x).group(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['gene'] = df['attributes'].apply(lambda x : re.search(r'gene_id\\s+\"(.+?)\";', x).group(1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference feature  start    end strand frame  \\\n",
       "0      ENm006    exon  71783  71788      -     .   \n",
       "1      ENm006     CDS  71783  71788      -     0   \n",
       "2      ENm006    exon  70312  70440      -     .   \n",
       "3      ENm006     CDS  70312  70440      -     0   \n",
       "4      ENm006    exon  69989  70210      -     .   \n",
       "..        ...     ...    ...    ...    ...   ...   \n",
       "405    ENm006     CDS  56303  56327      +     0   \n",
       "406    ENm006    exon  56689  57573      +     .   \n",
       "407    ENm006     CDS  56689  57573      +     2   \n",
       "408    ENm006    exon  57680  58323      +     .   \n",
       "409    ENm006     CDS  57680  57885      +     2   \n",
       "\n",
       "                                            attributes transcript_id     gene  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  U52112.4-005  ARHGAP4  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  U52112.4-005  ARHGAP4  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  U52112.4-005  ARHGAP4  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  U52112.4-005  ARHGAP4  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  U52112.4-005  ARHGAP4  \n",
       "..                                                 ...           ...      ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  U52112.2-001    AVPR2  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  U52112.2-001    AVPR2  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  U52112.2-001    AVPR2  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  U52112.2-001    AVPR2  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  U52112.2-001    AVPR2  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop('attributes', axis = 1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference feature  start    end strand frame transcript_id     gene\n",
       "0      ENm006    exon  71783  71788      -     .  U52112.4-005  ARHGAP4\n",
       "1      ENm006     CDS  71783  71788      -     0  U52112.4-005  ARHGAP4\n",
       "2      ENm006    exon  70312  70440      -     .  U52112.4-005  ARHGAP4\n",
       "3      ENm006     CDS  70312  70440      -     0  U52112.4-005  ARHGAP4\n",
       "4      ENm006    exon  69989  70210      -     .  U52112.4-005  ARHGAP4\n",
       "..        ...     ...    ...    ...    ...   ...           ...      ...\n",
       "405    ENm006     CDS  56303  56327      +     0  U52112.2-001    AVPR2\n",
       "406    ENm006    exon  56689  57573      +     .  U52112.2-001    AVPR2\n",
       "407    ENm006     CDS  56689  57573      +     2  U52112.2-001    AVPR2\n",
       "408    ENm006    exon  57680  58323      +     .  U52112.2-001    AVPR2\n",
       "409    ENm006     CDS  57680  57885      +     2  U52112.2-001    AVPR2\n",
       "\n",
       "[410 rows x 8 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Aggiungere la colonna `length` contenente la lunghezza della *feature*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['length'] = df['end'] - df['start'] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.4-005</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference feature  start    end strand frame transcript_id     gene  \\\n",
       "0      ENm006    exon  71783  71788      -     .  U52112.4-005  ARHGAP4   \n",
       "1      ENm006     CDS  71783  71788      -     0  U52112.4-005  ARHGAP4   \n",
       "2      ENm006    exon  70312  70440      -     .  U52112.4-005  ARHGAP4   \n",
       "3      ENm006     CDS  70312  70440      -     0  U52112.4-005  ARHGAP4   \n",
       "4      ENm006    exon  69989  70210      -     .  U52112.4-005  ARHGAP4   \n",
       "..        ...     ...    ...    ...    ...   ...           ...      ...   \n",
       "405    ENm006     CDS  56303  56327      +     0  U52112.2-001    AVPR2   \n",
       "406    ENm006    exon  56689  57573      +     .  U52112.2-001    AVPR2   \n",
       "407    ENm006     CDS  56689  57573      +     2  U52112.2-001    AVPR2   \n",
       "408    ENm006    exon  57680  58323      +     .  U52112.2-001    AVPR2   \n",
       "409    ENm006     CDS  57680  57885      +     2  U52112.2-001    AVPR2   \n",
       "\n",
       "     length  \n",
       "0         6  \n",
       "1         6  \n",
       "2       129  \n",
       "3       129  \n",
       "4       222  \n",
       "..      ...  \n",
       "405      25  \n",
       "406     885  \n",
       "407     885  \n",
       "408     644  \n",
       "409     206  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.reindex(columns = ['reference', 'feature', 'start', 'end',\n",
    "                      'length', 'strand', 'frame', 'gene', 'transcript_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>25</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>206</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference feature  start    end  length strand frame     gene  \\\n",
       "0      ENm006    exon  71783  71788       6      -     .  ARHGAP4   \n",
       "1      ENm006     CDS  71783  71788       6      -     0  ARHGAP4   \n",
       "2      ENm006    exon  70312  70440     129      -     .  ARHGAP4   \n",
       "3      ENm006     CDS  70312  70440     129      -     0  ARHGAP4   \n",
       "4      ENm006    exon  69989  70210     222      -     .  ARHGAP4   \n",
       "..        ...     ...    ...    ...     ...    ...   ...      ...   \n",
       "405    ENm006     CDS  56303  56327      25      +     0    AVPR2   \n",
       "406    ENm006    exon  56689  57573     885      +     .    AVPR2   \n",
       "407    ENm006     CDS  56689  57573     885      +     2    AVPR2   \n",
       "408    ENm006    exon  57680  58323     644      +     .    AVPR2   \n",
       "409    ENm006     CDS  57680  57885     206      +     2    AVPR2   \n",
       "\n",
       "    transcript_id  \n",
       "0    U52112.4-005  \n",
       "1    U52112.4-005  \n",
       "2    U52112.4-005  \n",
       "3    U52112.4-005  \n",
       "4    U52112.4-005  \n",
       "..            ...  \n",
       "405  U52112.2-001  \n",
       "406  U52112.2-001  \n",
       "407  U52112.2-001  \n",
       "408  U52112.2-001  \n",
       "409  U52112.2-001  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rimuovere tutte le *features* di lunghezza ≤ 6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['length'] > 6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>64935</td>\n",
       "      <td>65036</td>\n",
       "      <td>102</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>25</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>206</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference feature  start    end  length strand frame     gene  \\\n",
       "2      ENm006    exon  70312  70440     129      -     .  ARHGAP4   \n",
       "3      ENm006     CDS  70312  70440     129      -     0  ARHGAP4   \n",
       "4      ENm006    exon  69989  70210     222      -     .  ARHGAP4   \n",
       "5      ENm006     CDS  69989  70210     222      -     0  ARHGAP4   \n",
       "6      ENm006    exon  64935  65036     102      -     .  ARHGAP4   \n",
       "..        ...     ...    ...    ...     ...    ...   ...      ...   \n",
       "405    ENm006     CDS  56303  56327      25      +     0    AVPR2   \n",
       "406    ENm006    exon  56689  57573     885      +     .    AVPR2   \n",
       "407    ENm006     CDS  56689  57573     885      +     2    AVPR2   \n",
       "408    ENm006    exon  57680  58323     644      +     .    AVPR2   \n",
       "409    ENm006     CDS  57680  57885     206      +     2    AVPR2   \n",
       "\n",
       "    transcript_id  \n",
       "2    U52112.4-005  \n",
       "3    U52112.4-005  \n",
       "4    U52112.4-005  \n",
       "5    U52112.4-005  \n",
       "6    U52112.4-005  \n",
       "..            ...  \n",
       "405  U52112.2-001  \n",
       "406  U52112.2-001  \n",
       "407  U52112.2-001  \n",
       "408  U52112.2-001  \n",
       "409  U52112.2-001  \n",
       "\n",
       "[408 rows x 9 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estrarre il *data frame* dei 20 esoni più lunghi e degli ultimi 20 esoni localizzati sulla *reference sequence*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_df = df[df['feature'] == 'exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_df = exon_df.sort_values('length', ascending = False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "second_df = exon_df.sort_values('end').tail(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57938</td>\n",
       "      <td>1250</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57176</td>\n",
       "      <td>57573</td>\n",
       "      <td>398</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58322</td>\n",
       "      <td>643</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58524</td>\n",
       "      <td>59119</td>\n",
       "      <td>596</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58533</td>\n",
       "      <td>59119</td>\n",
       "      <td>587</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58534</td>\n",
       "      <td>59119</td>\n",
       "      <td>586</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58534</td>\n",
       "      <td>59119</td>\n",
       "      <td>586</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58572</td>\n",
       "      <td>59119</td>\n",
       "      <td>548</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58596</td>\n",
       "      <td>59119</td>\n",
       "      <td>524</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58611</td>\n",
       "      <td>59119</td>\n",
       "      <td>509</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58611</td>\n",
       "      <td>59119</td>\n",
       "      <td>509</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58611</td>\n",
       "      <td>59119</td>\n",
       "      <td>509</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>58626</td>\n",
       "      <td>59119</td>\n",
       "      <td>494</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>61169</td>\n",
       "      <td>61676</td>\n",
       "      <td>508</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>61328</td>\n",
       "      <td>61768</td>\n",
       "      <td>441</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>64935</td>\n",
       "      <td>65410</td>\n",
       "      <td>476</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>71569</td>\n",
       "      <td>71965</td>\n",
       "      <td>397</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>543097</td>\n",
       "      <td>545706</td>\n",
       "      <td>2610</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546425</td>\n",
       "      <td>111</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546508</td>\n",
       "      <td>194</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546508</td>\n",
       "      <td>194</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546508</td>\n",
       "      <td>194</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>546980</td>\n",
       "      <td>547020</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>546980</td>\n",
       "      <td>547020</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>546980</td>\n",
       "      <td>547020</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>547684</td>\n",
       "      <td>547769</td>\n",
       "      <td>86</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>547684</td>\n",
       "      <td>547769</td>\n",
       "      <td>86</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>547684</td>\n",
       "      <td>547769</td>\n",
       "      <td>86</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>547709</td>\n",
       "      <td>547769</td>\n",
       "      <td>61</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-006</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   reference feature   start     end  length strand frame     gene  \\\n",
       "0     ENm006    exon   56689   57573     885      +     .    AVPR2   \n",
       "1     ENm006    exon   56689   57938    1250      +     .    AVPR2   \n",
       "2     ENm006    exon   57176   57573     398      +     .    AVPR2   \n",
       "3     ENm006    exon   57680   58322     643      +     .    AVPR2   \n",
       "4     ENm006    exon   57680   58323     644      +     .    AVPR2   \n",
       "5     ENm006    exon   58524   59119     596      -     .  ARHGAP4   \n",
       "6     ENm006    exon   58533   59119     587      -     .  ARHGAP4   \n",
       "7     ENm006    exon   58534   59119     586      -     .  ARHGAP4   \n",
       "8     ENm006    exon   58534   59119     586      -     .  ARHGAP4   \n",
       "9     ENm006    exon   58572   59119     548      -     .  ARHGAP4   \n",
       "10    ENm006    exon   58596   59119     524      -     .  ARHGAP4   \n",
       "11    ENm006    exon   58611   59119     509      -     .  ARHGAP4   \n",
       "12    ENm006    exon   58611   59119     509      -     .  ARHGAP4   \n",
       "13    ENm006    exon   58611   59119     509      -     .  ARHGAP4   \n",
       "14    ENm006    exon   58626   59119     494      -     .  ARHGAP4   \n",
       "15    ENm006    exon   61169   61676     508      -     .  ARHGAP4   \n",
       "16    ENm006    exon   61328   61768     441      -     .  ARHGAP4   \n",
       "17    ENm006    exon   64935   65410     476      -     .  ARHGAP4   \n",
       "18    ENm006    exon   71569   71965     397      -     .  ARHGAP4   \n",
       "19    ENm006    exon  543097  545706    2610      +     .  ATP6AP1   \n",
       "20    ENm006    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "21    ENm006    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "22    ENm006    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "23    ENm006    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "24    ENm006    exon  546315  546425     111      +     .  ATP6AP1   \n",
       "25    ENm006    exon  546315  546508     194      +     .  ATP6AP1   \n",
       "26    ENm006    exon  546315  546508     194      +     .  ATP6AP1   \n",
       "27    ENm006    exon  546315  546508     194      +     .  ATP6AP1   \n",
       "28    ENm006    exon  546980  547020      41      +     .  ATP6AP1   \n",
       "29    ENm006    exon  546980  547020      41      +     .  ATP6AP1   \n",
       "30    ENm006    exon  546980  547020      41      +     .  ATP6AP1   \n",
       "31    ENm006    exon  547684  547769      86      +     .  ATP6AP1   \n",
       "32    ENm006    exon  547684  547769      86      +     .  ATP6AP1   \n",
       "33    ENm006    exon  547684  547769      86      +     .  ATP6AP1   \n",
       "34    ENm006    exon  547709  547769      61      +     .  ATP6AP1   \n",
       "35    ENm006    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "36    ENm006    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "37    ENm006    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "38    ENm006    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "39    ENm006    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "\n",
       "         transcript_id  \n",
       "0         U52112.2-001  \n",
       "1         U52112.2-003  \n",
       "2         U52112.2-002  \n",
       "3         U52112.2-002  \n",
       "4         U52112.2-001  \n",
       "5         U52112.4-024  \n",
       "6         U52112.4-001  \n",
       "7         U52112.4-002  \n",
       "8         U52112.4-003  \n",
       "9         U52112.4-016  \n",
       "10        U52112.4-006  \n",
       "11        U52112.4-011  \n",
       "12        U52112.4-012  \n",
       "13        U52112.4-013  \n",
       "14        U52112.4-014  \n",
       "15        U52112.4-007  \n",
       "16        U52112.4-016  \n",
       "17        U52112.4-023  \n",
       "18        U52112.4-021  \n",
       "19  XX-FW83563B9.4-002  \n",
       "20  XX-FW83563B9.4-001  \n",
       "21  XX-FW83563B9.4-003  \n",
       "22  XX-FW83563B9.4-004  \n",
       "23  XX-FW83563B9.4-006  \n",
       "24  XX-FW83563B9.4-003  \n",
       "25  XX-FW83563B9.4-001  \n",
       "26  XX-FW83563B9.4-002  \n",
       "27  XX-FW83563B9.4-004  \n",
       "28  XX-FW83563B9.4-001  \n",
       "29  XX-FW83563B9.4-002  \n",
       "30  XX-FW83563B9.4-004  \n",
       "31  XX-FW83563B9.4-001  \n",
       "32  XX-FW83563B9.4-002  \n",
       "33  XX-FW83563B9.4-004  \n",
       "34  XX-FW83563B9.4-006  \n",
       "35  XX-FW83563B9.4-001  \n",
       "36  XX-FW83563B9.4-002  \n",
       "37  XX-FW83563B9.4-003  \n",
       "38  XX-FW83563B9.4-004  \n",
       "39  XX-FW83563B9.4-006  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(first_df, second_df, how = 'outer')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estrarre i nomi dei geni e gli identificatori dei trascritti annotati nel file GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determinare, per ogni gene, la lunghezza media, massima e minima degli esoni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determinare la lunghezza minima degli esoni ed estrarre tutti i trascritti che contengono un esone di lunghezza minima\n",
    "\n",
    "a) Determinare la lunghezza minima degli esoni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) Estrarre la lista dei trascritti che contengono un esone di lunghezza minima"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contare quanti trascritti sono annotati per il gene `ARHGAP4`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estrarre la lista dei geni con strand `+`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estrarre gli esoni (distinti) del gene `ATP6AP1` in una lista di tuple (start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Contare il numero di trascritti del gene `ARHGAP4` che hanno una CDS annotata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estrarre lo strand del gene  `ATP6AP1`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determinare il trascritto che ha più esoni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Estrarre per ogni trascritto del gene  `ARHGAP4` la lista delle tuple (start, end) dei suoi esoni ordinate per start crescente"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determinare, per ogni gene e ogni trascritto il numero di esoni che lo compongono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determinare per ogni gene, la lunghezza del suo *locus*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
