{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manipolazione di un file GTF (Gene Transfer Format) attraverso la libreria `Pandas`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1) Importare `Pandas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2) Leggere il file GTF\n",
    "\n",
    "    df = pd.read_csv(gtf_file_name, sep='\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('./input.gtf', sep='\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1     2      3      4  5  6  7  \\\n",
       "0    ENm006  VEGA_Known  exon  71783  71788  .  -  .   \n",
       "1    ENm006  VEGA_Known   CDS  71783  71788  .  -  0   \n",
       "2    ENm006  VEGA_Known  exon  70312  70440  .  -  .   \n",
       "3    ENm006  VEGA_Known   CDS  70312  70440  .  -  0   \n",
       "4    ENm006  VEGA_Known  exon  69989  70210  .  -  .   \n",
       "..      ...         ...   ...    ...    ... .. .. ..   \n",
       "405  ENm006  VEGA_Known   CDS  56303  56327  .  +  0   \n",
       "406  ENm006  VEGA_Known  exon  56689  57573  .  +  .   \n",
       "407  ENm006  VEGA_Known   CDS  56689  57573  .  +  2   \n",
       "408  ENm006  VEGA_Known  exon  57680  58323  .  +  .   \n",
       "409  ENm006  VEGA_Known   CDS  57680  57885  .  +  2   \n",
       "\n",
       "                                                     8  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NB**: `read_csv()` ha un attributo `names` che permette di specificare la lista dei nomi delle colonne del data frame restituito."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3) Cambiare i nomi delle colonne\n",
    "\n",
    "I nomi delle colonne devono essere:\n",
    "- reference\n",
    "- source\n",
    "- feature\n",
    "- start\n",
    "- end\n",
    "- score\n",
    "- strand\n",
    "- frame\n",
    "- attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "replace_dict = {0 : 'reference', 1 : 'source', 2 : 'feature', 3 : 'start', 4 : 'end', 5 : 'score', 6 : 'strand', 7 : 'frame', 8 : 'attributes'}\n",
    "df.rename(columns = replace_dict, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference      source feature  start    end score strand frame  \\\n",
       "0      ENm006  VEGA_Known    exon  71783  71788     .      -     .   \n",
       "1      ENm006  VEGA_Known     CDS  71783  71788     .      -     0   \n",
       "2      ENm006  VEGA_Known    exon  70312  70440     .      -     .   \n",
       "3      ENm006  VEGA_Known     CDS  70312  70440     .      -     0   \n",
       "4      ENm006  VEGA_Known    exon  69989  70210     .      -     .   \n",
       "..        ...         ...     ...    ...    ...   ...    ...   ...   \n",
       "405    ENm006  VEGA_Known     CDS  56303  56327     .      +     0   \n",
       "406    ENm006  VEGA_Known    exon  56689  57573     .      +     .   \n",
       "407    ENm006  VEGA_Known     CDS  56689  57573     .      +     2   \n",
       "408    ENm006  VEGA_Known    exon  57680  58323     .      +     .   \n",
       "409    ENm006  VEGA_Known     CDS  57680  57885     .      +     2   \n",
       "\n",
       "                                            attributes  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4) Eliminare le colonne `source` e `score` e sostituire l'identificatore `ENm006` con l'identificatore `ENCODE_REGION` in tutti i campi della colonna `reference`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(['source', 'score'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['reference'] = 'ENCODE_REGION'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end strand frame  \\\n",
       "0    ENCODE_REGION    exon  71783  71788      -     .   \n",
       "1    ENCODE_REGION     CDS  71783  71788      -     0   \n",
       "2    ENCODE_REGION    exon  70312  70440      -     .   \n",
       "3    ENCODE_REGION     CDS  70312  70440      -     0   \n",
       "4    ENCODE_REGION    exon  69989  70210      -     .   \n",
       "..             ...     ...    ...    ...    ...   ...   \n",
       "405  ENCODE_REGION     CDS  56303  56327      +     0   \n",
       "406  ENCODE_REGION    exon  56689  57573      +     .   \n",
       "407  ENCODE_REGION     CDS  56689  57573      +     2   \n",
       "408  ENCODE_REGION    exon  57680  58323      +     .   \n",
       "409  ENCODE_REGION     CDS  57680  57885      +     2   \n",
       "\n",
       "                                            attributes  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5) Sostituire la colonna degli attributi con le due colonne  `transcript` e `gene`\n",
    "\n",
    "La colonne `transcript` e `gene` dovranno contenere solo l'ID del trascritto e del gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['gene'] = ''\n",
    "df['transcript'] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "for (index, record) in df.iterrows():\n",
    "    transcript_id = re.search('transcript_id\\s+(.+?);', record['attributes']).group(1).replace('\"', '')\n",
    "    gene_id = re.search('gene_id\\s+(.+?);', record['attributes']).group(1).replace('\"', '')\n",
    "    df.loc[index, 'transcript'] =  transcript_id\n",
    "    df.loc[index, 'gene'] =  gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop('attributes', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end strand frame     gene    transcript\n",
       "0    ENCODE_REGION    exon  71783  71788      -     .  ARHGAP4  U52112.4-005\n",
       "1    ENCODE_REGION     CDS  71783  71788      -     0  ARHGAP4  U52112.4-005\n",
       "2    ENCODE_REGION    exon  70312  70440      -     .  ARHGAP4  U52112.4-005\n",
       "3    ENCODE_REGION     CDS  70312  70440      -     0  ARHGAP4  U52112.4-005\n",
       "4    ENCODE_REGION    exon  69989  70210      -     .  ARHGAP4  U52112.4-005\n",
       "..             ...     ...    ...    ...    ...   ...      ...           ...\n",
       "405  ENCODE_REGION     CDS  56303  56327      +     0    AVPR2  U52112.2-001\n",
       "406  ENCODE_REGION    exon  56689  57573      +     .    AVPR2  U52112.2-001\n",
       "407  ENCODE_REGION     CDS  56689  57573      +     2    AVPR2  U52112.2-001\n",
       "408  ENCODE_REGION    exon  57680  58323      +     .    AVPR2  U52112.2-001\n",
       "409  ENCODE_REGION     CDS  57680  57885      +     2    AVPR2  U52112.2-001\n",
       "\n",
       "[410 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 6) Aggiungere la colonna `length` contenente la lunghezza della feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['length'] = df['end'] - df['start'] + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reindicizzazione delle colonne:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.reindex(columns = ['reference', 'feature', 'start', 'end', 'length', 'strand', 'frame', 'gene', 'transcript'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>25</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>206</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end  length strand frame     gene  \\\n",
       "0    ENCODE_REGION    exon  71783  71788       6      -     .  ARHGAP4   \n",
       "1    ENCODE_REGION     CDS  71783  71788       6      -     0  ARHGAP4   \n",
       "2    ENCODE_REGION    exon  70312  70440     129      -     .  ARHGAP4   \n",
       "3    ENCODE_REGION     CDS  70312  70440     129      -     0  ARHGAP4   \n",
       "4    ENCODE_REGION    exon  69989  70210     222      -     .  ARHGAP4   \n",
       "..             ...     ...    ...    ...     ...    ...   ...      ...   \n",
       "405  ENCODE_REGION     CDS  56303  56327      25      +     0    AVPR2   \n",
       "406  ENCODE_REGION    exon  56689  57573     885      +     .    AVPR2   \n",
       "407  ENCODE_REGION     CDS  56689  57573     885      +     2    AVPR2   \n",
       "408  ENCODE_REGION    exon  57680  58323     644      +     .    AVPR2   \n",
       "409  ENCODE_REGION     CDS  57680  57885     206      +     2    AVPR2   \n",
       "\n",
       "       transcript  \n",
       "0    U52112.4-005  \n",
       "1    U52112.4-005  \n",
       "2    U52112.4-005  \n",
       "3    U52112.4-005  \n",
       "4    U52112.4-005  \n",
       "..            ...  \n",
       "405  U52112.2-001  \n",
       "406  U52112.2-001  \n",
       "407  U52112.2-001  \n",
       "408  U52112.2-001  \n",
       "409  U52112.2-001  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 7) Rimuovere dal *data frame* tutte le features di lunghezza minore o uguale a 6 basi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>64935</td>\n",
       "      <td>65036</td>\n",
       "      <td>102</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>25</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>206</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end  length strand frame     gene  \\\n",
       "2    ENCODE_REGION    exon  70312  70440     129      -     .  ARHGAP4   \n",
       "3    ENCODE_REGION     CDS  70312  70440     129      -     0  ARHGAP4   \n",
       "4    ENCODE_REGION    exon  69989  70210     222      -     .  ARHGAP4   \n",
       "5    ENCODE_REGION     CDS  69989  70210     222      -     0  ARHGAP4   \n",
       "6    ENCODE_REGION    exon  64935  65036     102      -     .  ARHGAP4   \n",
       "..             ...     ...    ...    ...     ...    ...   ...      ...   \n",
       "405  ENCODE_REGION     CDS  56303  56327      25      +     0    AVPR2   \n",
       "406  ENCODE_REGION    exon  56689  57573     885      +     .    AVPR2   \n",
       "407  ENCODE_REGION     CDS  56689  57573     885      +     2    AVPR2   \n",
       "408  ENCODE_REGION    exon  57680  58323     644      +     .    AVPR2   \n",
       "409  ENCODE_REGION     CDS  57680  57885     206      +     2    AVPR2   \n",
       "\n",
       "       transcript  \n",
       "2    U52112.4-005  \n",
       "3    U52112.4-005  \n",
       "4    U52112.4-005  \n",
       "5    U52112.4-005  \n",
       "6    U52112.4-005  \n",
       "..            ...  \n",
       "405  U52112.2-001  \n",
       "406  U52112.2-001  \n",
       "407  U52112.2-001  \n",
       "408  U52112.2-001  \n",
       "409  U52112.2-001  \n",
       "\n",
       "[408 rows x 9 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = df['length'] > 6\n",
    "df[mask]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In alternativa, con il metodo `drop()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = df['length'] <= 6\n",
    "df.drop(df[mask].index, axis = 0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>64935</td>\n",
       "      <td>65036</td>\n",
       "      <td>102</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>25</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>206</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>408 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end  length strand frame     gene  \\\n",
       "2    ENCODE_REGION    exon  70312  70440     129      -     .  ARHGAP4   \n",
       "3    ENCODE_REGION     CDS  70312  70440     129      -     0  ARHGAP4   \n",
       "4    ENCODE_REGION    exon  69989  70210     222      -     .  ARHGAP4   \n",
       "5    ENCODE_REGION     CDS  69989  70210     222      -     0  ARHGAP4   \n",
       "6    ENCODE_REGION    exon  64935  65036     102      -     .  ARHGAP4   \n",
       "..             ...     ...    ...    ...     ...    ...   ...      ...   \n",
       "405  ENCODE_REGION     CDS  56303  56327      25      +     0    AVPR2   \n",
       "406  ENCODE_REGION    exon  56689  57573     885      +     .    AVPR2   \n",
       "407  ENCODE_REGION     CDS  56689  57573     885      +     2    AVPR2   \n",
       "408  ENCODE_REGION    exon  57680  58323     644      +     .    AVPR2   \n",
       "409  ENCODE_REGION     CDS  57680  57885     206      +     2    AVPR2   \n",
       "\n",
       "       transcript  \n",
       "2    U52112.4-005  \n",
       "3    U52112.4-005  \n",
       "4    U52112.4-005  \n",
       "5    U52112.4-005  \n",
       "6    U52112.4-005  \n",
       "..            ...  \n",
       "405  U52112.2-001  \n",
       "406  U52112.2-001  \n",
       "407  U52112.2-001  \n",
       "408  U52112.2-001  \n",
       "409  U52112.2-001  \n",
       "\n",
       "[408 rows x 9 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 8) Estrarre il *data frame* dei primi 20 esoni più lunghi e degli ultimi 20 esoni localizzati sul *reference*\n",
    "\n",
    "*SUGGERIMENTO*: usare la funzione:\n",
    "\n",
    "    pd.merge(first_df, second_df, how='outer')\n",
    "    \n",
    "per unire verticalmente due *data frame* con le stesse colonne."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "esoni_df = df[df.feature == 'exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "first_df = esoni_df.sort_values('length', ascending = False, inplace = False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "second_df = esoni_df.sort_values('start', ascending = False, inplace = False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>543097</td>\n",
       "      <td>545706</td>\n",
       "      <td>2610</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57938</td>\n",
       "      <td>1250</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58322</td>\n",
       "      <td>643</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58524</td>\n",
       "      <td>59119</td>\n",
       "      <td>596</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58533</td>\n",
       "      <td>59119</td>\n",
       "      <td>587</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58534</td>\n",
       "      <td>59119</td>\n",
       "      <td>586</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58534</td>\n",
       "      <td>59119</td>\n",
       "      <td>586</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58572</td>\n",
       "      <td>59119</td>\n",
       "      <td>548</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58596</td>\n",
       "      <td>59119</td>\n",
       "      <td>524</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58611</td>\n",
       "      <td>59119</td>\n",
       "      <td>509</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58611</td>\n",
       "      <td>59119</td>\n",
       "      <td>509</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58611</td>\n",
       "      <td>59119</td>\n",
       "      <td>509</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>61169</td>\n",
       "      <td>61676</td>\n",
       "      <td>508</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>58626</td>\n",
       "      <td>59119</td>\n",
       "      <td>494</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>64935</td>\n",
       "      <td>65410</td>\n",
       "      <td>476</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>61328</td>\n",
       "      <td>61768</td>\n",
       "      <td>441</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57176</td>\n",
       "      <td>57573</td>\n",
       "      <td>398</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>71569</td>\n",
       "      <td>71965</td>\n",
       "      <td>397</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>548257</td>\n",
       "      <td>548495</td>\n",
       "      <td>239</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>547709</td>\n",
       "      <td>547769</td>\n",
       "      <td>61</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>547684</td>\n",
       "      <td>547769</td>\n",
       "      <td>86</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>547684</td>\n",
       "      <td>547769</td>\n",
       "      <td>86</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>547684</td>\n",
       "      <td>547769</td>\n",
       "      <td>86</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>546980</td>\n",
       "      <td>547020</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>546980</td>\n",
       "      <td>547020</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>546980</td>\n",
       "      <td>547020</td>\n",
       "      <td>41</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546508</td>\n",
       "      <td>194</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546508</td>\n",
       "      <td>194</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546425</td>\n",
       "      <td>111</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>546315</td>\n",
       "      <td>546508</td>\n",
       "      <td>194</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>545879</td>\n",
       "      <td>545953</td>\n",
       "      <td>75</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ATP6AP1</td>\n",
       "      <td>XX-FW83563B9.4-002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        reference feature   start     end  length strand frame     gene  \\\n",
       "0   ENCODE_REGION    exon  543097  545706    2610      +     .  ATP6AP1   \n",
       "1   ENCODE_REGION    exon   56689   57938    1250      +     .    AVPR2   \n",
       "2   ENCODE_REGION    exon   56689   57573     885      +     .    AVPR2   \n",
       "3   ENCODE_REGION    exon   57680   58323     644      +     .    AVPR2   \n",
       "4   ENCODE_REGION    exon   57680   58322     643      +     .    AVPR2   \n",
       "5   ENCODE_REGION    exon   58524   59119     596      -     .  ARHGAP4   \n",
       "6   ENCODE_REGION    exon   58533   59119     587      -     .  ARHGAP4   \n",
       "7   ENCODE_REGION    exon   58534   59119     586      -     .  ARHGAP4   \n",
       "8   ENCODE_REGION    exon   58534   59119     586      -     .  ARHGAP4   \n",
       "9   ENCODE_REGION    exon   58572   59119     548      -     .  ARHGAP4   \n",
       "10  ENCODE_REGION    exon   58596   59119     524      -     .  ARHGAP4   \n",
       "11  ENCODE_REGION    exon   58611   59119     509      -     .  ARHGAP4   \n",
       "12  ENCODE_REGION    exon   58611   59119     509      -     .  ARHGAP4   \n",
       "13  ENCODE_REGION    exon   58611   59119     509      -     .  ARHGAP4   \n",
       "14  ENCODE_REGION    exon   61169   61676     508      -     .  ARHGAP4   \n",
       "15  ENCODE_REGION    exon   58626   59119     494      -     .  ARHGAP4   \n",
       "16  ENCODE_REGION    exon   64935   65410     476      -     .  ARHGAP4   \n",
       "17  ENCODE_REGION    exon   61328   61768     441      -     .  ARHGAP4   \n",
       "18  ENCODE_REGION    exon   57176   57573     398      +     .    AVPR2   \n",
       "19  ENCODE_REGION    exon   71569   71965     397      -     .  ARHGAP4   \n",
       "20  ENCODE_REGION    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "21  ENCODE_REGION    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "22  ENCODE_REGION    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "23  ENCODE_REGION    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "24  ENCODE_REGION    exon  548257  548495     239      +     .  ATP6AP1   \n",
       "25  ENCODE_REGION    exon  547709  547769      61      +     .  ATP6AP1   \n",
       "26  ENCODE_REGION    exon  547684  547769      86      +     .  ATP6AP1   \n",
       "27  ENCODE_REGION    exon  547684  547769      86      +     .  ATP6AP1   \n",
       "28  ENCODE_REGION    exon  547684  547769      86      +     .  ATP6AP1   \n",
       "29  ENCODE_REGION    exon  546980  547020      41      +     .  ATP6AP1   \n",
       "30  ENCODE_REGION    exon  546980  547020      41      +     .  ATP6AP1   \n",
       "31  ENCODE_REGION    exon  546980  547020      41      +     .  ATP6AP1   \n",
       "32  ENCODE_REGION    exon  546315  546508     194      +     .  ATP6AP1   \n",
       "33  ENCODE_REGION    exon  546315  546508     194      +     .  ATP6AP1   \n",
       "34  ENCODE_REGION    exon  546315  546425     111      +     .  ATP6AP1   \n",
       "35  ENCODE_REGION    exon  546315  546508     194      +     .  ATP6AP1   \n",
       "36  ENCODE_REGION    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "37  ENCODE_REGION    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "38  ENCODE_REGION    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "39  ENCODE_REGION    exon  545879  545953      75      +     .  ATP6AP1   \n",
       "\n",
       "            transcript  \n",
       "0   XX-FW83563B9.4-002  \n",
       "1         U52112.2-003  \n",
       "2         U52112.2-001  \n",
       "3         U52112.2-001  \n",
       "4         U52112.2-002  \n",
       "5         U52112.4-024  \n",
       "6         U52112.4-001  \n",
       "7         U52112.4-003  \n",
       "8         U52112.4-002  \n",
       "9         U52112.4-016  \n",
       "10        U52112.4-006  \n",
       "11        U52112.4-011  \n",
       "12        U52112.4-013  \n",
       "13        U52112.4-012  \n",
       "14        U52112.4-007  \n",
       "15        U52112.4-014  \n",
       "16        U52112.4-023  \n",
       "17        U52112.4-016  \n",
       "18        U52112.2-002  \n",
       "19        U52112.4-021  \n",
       "20  XX-FW83563B9.4-006  \n",
       "21  XX-FW83563B9.4-004  \n",
       "22  XX-FW83563B9.4-001  \n",
       "23  XX-FW83563B9.4-003  \n",
       "24  XX-FW83563B9.4-002  \n",
       "25  XX-FW83563B9.4-006  \n",
       "26  XX-FW83563B9.4-004  \n",
       "27  XX-FW83563B9.4-001  \n",
       "28  XX-FW83563B9.4-002  \n",
       "29  XX-FW83563B9.4-004  \n",
       "30  XX-FW83563B9.4-001  \n",
       "31  XX-FW83563B9.4-002  \n",
       "32  XX-FW83563B9.4-004  \n",
       "33  XX-FW83563B9.4-001  \n",
       "34  XX-FW83563B9.4-003  \n",
       "35  XX-FW83563B9.4-002  \n",
       "36  XX-FW83563B9.4-004  \n",
       "37  XX-FW83563B9.4-001  \n",
       "38  XX-FW83563B9.4-003  \n",
       "39  XX-FW83563B9.4-002  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(first_df, second_df, how='outer')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 9) Estrarre la lista dei geni presenti nel file GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ARHGAP4', 'ATP6AP1', 'AVPR2']"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df['gene'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 10) Estrarre la lista dei trascritti presenti nel file GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['U52112.4-005',\n",
       " 'U52112.4-018',\n",
       " 'U52112.4-014',\n",
       " 'U52112.4-022',\n",
       " 'U52112.4-021',\n",
       " 'U52112.4-006',\n",
       " 'U52112.4-019',\n",
       " 'U52112.4-017',\n",
       " 'U52112.4-013',\n",
       " 'U52112.4-002',\n",
       " 'U52112.4-016',\n",
       " 'U52112.4-010',\n",
       " 'U52112.4-009',\n",
       " 'U52112.4-020',\n",
       " 'U52112.4-003',\n",
       " 'U52112.4-012',\n",
       " 'U52112.4-001',\n",
       " 'U52112.4-024',\n",
       " 'U52112.4-007',\n",
       " 'U52112.4-015',\n",
       " 'U52112.4-008',\n",
       " 'U52112.4-004',\n",
       " 'U52112.4-011',\n",
       " 'U52112.4-023',\n",
       " 'XX-FW83563B9.4-002',\n",
       " 'XX-FW83563B9.4-003',\n",
       " 'XX-FW83563B9.4-004',\n",
       " 'XX-FW83563B9.4-006',\n",
       " 'XX-FW83563B9.4-001',\n",
       " 'U52112.2-003',\n",
       " 'U52112.2-002',\n",
       " 'U52112.2-001']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df['transcript'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 11) Determinare, per ogni gene, la lunghezza media, massima e minima degli esoni (considerando anche gli esoni duplicati dovuti al fatto che un esone può appartenere a trascritti diversi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = df.feature == 'exon'\n",
    "gb = df[mask].groupby('gene')['length']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "ARHGAP4    149.16996\n",
       "ATP6AP1    224.00000\n",
       "AVPR2      467.00000\n",
       "Name: length, dtype: float64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "ARHGAP4    28\n",
       "ATP6AP1    41\n",
       "AVPR2      37\n",
       "Name: length, dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "ARHGAP4     596\n",
       "ATP6AP1    2610\n",
       "AVPR2      1250\n",
       "Name: length, dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb.max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 12) Determinare la lunghezza minima degli esoni ed estrarre tutti i trascritti che contengono un esone di lunghezza minima"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_length = df[df.feature == 'exon'].length.min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = (df.feature == 'exon') & (df.length == min_length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['U52112.4-018',\n",
       " 'U52112.4-014',\n",
       " 'U52112.4-013',\n",
       " 'U52112.4-002',\n",
       " 'U52112.4-003',\n",
       " 'U52112.4-012',\n",
       " 'U52112.4-001',\n",
       " 'U52112.4-024',\n",
       " 'U52112.4-008',\n",
       " 'U52112.4-011']"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df[mask]['transcript'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 13) Contare quanti trascritti sono annotati per il gene `ARHGAP4`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "24"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mask = (df.feature == 'exon') & (df.gene == 'ARHGAP4')\n",
    "len(df[mask]['transcript'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 14) Estrarre la lista dei geni con strand `+`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ATP6AP1', 'AVPR2']"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(df[df.strand == '+']['gene'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 15) Estrarre gli esoni (distinti) del gene `ATP6AP1` in una lista di tuple (start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_temp = df[(df.gene == 'ATP6AP1') & (df.feature == 'exon')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_temp = df_temp[['start', 'end']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(542747, 542902),\n",
       " (543097, 545706),\n",
       " (545879, 545953),\n",
       " (546315, 546508),\n",
       " (546980, 547020),\n",
       " (547684, 547769),\n",
       " (548257, 548495),\n",
       " (542694, 542902),\n",
       " (543097, 543223),\n",
       " (546315, 546425),\n",
       " (542894, 543223),\n",
       " (542790, 542902),\n",
       " (547709, 547769),\n",
       " (542687, 542902)]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(zip(df_temp['start'], df_temp['end']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 16) Contare i trascritti del gene `ARHGAP4` che hanno una CDS annotata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df[(df.gene == 'ARHGAP4') & (df.feature == 'CDS')]['transcript'].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 17) Estrarre lo strand del gene  `ATP6AP1`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'+'"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.gene == 'ATP6AP1']['strand'].unique()[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 18) Determinare il trascritto che ha più esoni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_df = df[df.feature == 'exon']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "23"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_df.groupby('transcript')['feature'].count().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['U52112.4-005',\n",
       " 'U52112.4-018',\n",
       " 'U52112.4-014',\n",
       " 'U52112.4-022',\n",
       " 'U52112.4-021',\n",
       " 'U52112.4-006',\n",
       " 'U52112.4-019',\n",
       " 'U52112.4-017',\n",
       " 'U52112.4-013',\n",
       " 'U52112.4-002',\n",
       " 'U52112.4-016',\n",
       " 'U52112.4-010',\n",
       " 'U52112.4-009',\n",
       " 'U52112.4-020',\n",
       " 'U52112.4-003',\n",
       " 'U52112.4-012',\n",
       " 'U52112.4-001',\n",
       " 'U52112.4-024',\n",
       " 'U52112.4-007',\n",
       " 'U52112.4-015',\n",
       " 'U52112.4-008',\n",
       " 'U52112.4-004',\n",
       " 'U52112.4-011',\n",
       " 'U52112.4-023',\n",
       " 'XX-FW83563B9.4-002',\n",
       " 'XX-FW83563B9.4-003',\n",
       " 'XX-FW83563B9.4-004',\n",
       " 'XX-FW83563B9.4-006',\n",
       " 'XX-FW83563B9.4-001',\n",
       " 'U52112.2-003',\n",
       " 'U52112.2-002',\n",
       " 'U52112.2-001']"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_of_transcripts = list(exon_df['transcript'].unique())\n",
    "list_of_transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_value = 0\n",
    "res = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "for transcript in list_of_transcripts:\n",
    "    value = exon_df[exon_df.transcript == transcript]['feature'].count()\n",
    "    if value > max_value:\n",
    "        max_value = value\n",
    "        res = transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'U52112.4-001'"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In alternativa:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "groups_by_transcript = exon_df.groupby('transcript').groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'U52112.4-001'"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(map(lambda x: (len(x[1]), x[0]), groups_by_transcript.items()), reverse = True).pop(0)[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 19) Estrarre per ogni trascritto del gene  `ARHGAP4` la lista delle tuple (start, end) dei suoi esoni ordinate per start crescente"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = (df.feature == 'exon') & (df.gene == 'ARHGAP4')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [],
   "source": [
    "groups_by_transcript = df[mask].groupby('transcript').groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "U52112.4-001\n",
      "[(58533, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64459), (64566, 64757), (64935, 65036), (69989, 70210), (70312, 70440), (70724, 70843), (71783, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (77293, 77417)]\n",
      "U52112.4-002\n",
      "[(58534, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61857, 61991), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64459), (64566, 64757), (64935, 65036), (69989, 70210), (70312, 70440), (71783, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (77293, 77401)]\n",
      "U52112.4-003\n",
      "[(58534, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64459), (64566, 64757), (64935, 65036), (69989, 70210), (70312, 70440), (71783, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (77293, 77401)]\n",
      "U52112.4-004\n",
      "[(70119, 70210), (70312, 70440), (71783, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (73280, 73404)]\n",
      "U52112.4-005\n",
      "[(64385, 64459), (64566, 64673), (64935, 65036), (69989, 70210), (70312, 70440)]\n",
      "U52112.4-006\n",
      "[(58596, 59119), (60227, 60326), (60600, 60815)]\n",
      "U52112.4-007\n",
      "[(60976, 61081), (61169, 61676)]\n",
      "U52112.4-008\n",
      "[(60665, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 62028)]\n",
      "U52112.4-009\n",
      "[(60625, 60692), (60898, 61081), (61169, 61242), (61663, 61768), (61857, 61991), (62079, 62258)]\n",
      "U52112.4-010\n",
      "[(60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62286, 62346), (63857, 63942)]\n",
      "U52112.4-011\n",
      "[(58611, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64459), (64566, 64757), (64935, 65036), (69989, 70210), (70312, 70440), (71783, 71965), (72521, 72683), (72761, 72965), (77293, 77401)]\n",
      "U52112.4-012\n",
      "[(58611, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64757), (64935, 65036), (69989, 70210), (72521, 72683), (72761, 72965), (77293, 77401)]\n",
      "U52112.4-013\n",
      "[(58611, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64757), (64935, 65036), (72521, 72683), (72761, 72965), (77293, 77401)]\n",
      "U52112.4-014\n",
      "[(58626, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64459), (64566, 64757), (64935, 65036), (69989, 70210), (70312, 70440), (71783, 71993), (72253, 72315), (72521, 72683), (72761, 72965), (77293, 77462)]\n",
      "U52112.4-015\n",
      "[(69999, 70210), (70312, 70440), (71783, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (77747, 77914)]\n",
      "U52112.4-016\n",
      "[(58572, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63894)]\n",
      "U52112.4-017\n",
      "[(71865, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (79484, 79576)]\n",
      "U52112.4-018\n",
      "[(71896, 71965), (72253, 72379), (72521, 72683), (72761, 72965), (79484, 79511)]\n",
      "U52112.4-019\n",
      "[(71872, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (77973, 78017)]\n",
      "U52112.4-020\n",
      "[(70097, 70210), (70312, 70440), (70724, 70843), (71783, 71965), (72521, 72560)]\n",
      "U52112.4-021\n",
      "[(71569, 71965), (72253, 72315), (72521, 72556)]\n",
      "U52112.4-022\n",
      "[(64367, 64757), (64935, 65036), (69989, 70210), (70312, 70440), (71783, 71965), (72253, 72315), (72521, 72683), (72761, 72965), (83227, 83271), (83472, 83587), (83695, 83740), (85099, 85157), (85533, 85631), (86040, 86155)]\n",
      "U52112.4-023\n",
      "[(64681, 64757), (64935, 65410)]\n",
      "U52112.4-024\n",
      "[(58524, 59119), (60227, 60326), (60600, 60692), (60898, 61081), (61169, 61242), (61328, 61561), (61663, 61768), (61857, 61991), (62079, 62156), (62286, 62346), (63857, 63959), (64181, 64208), (64375, 64459), (64566, 64757), (64935, 65036), (72253, 72315), (72521, 72683), (72761, 72965), (77293, 77411)]\n"
     ]
    }
   ],
   "source": [
    "for transcript_id in groups_by_transcript:\n",
    "    print(transcript_id)\n",
    "    index_list = groups_by_transcript[transcript_id]\n",
    "    \n",
    "    start_list = df.loc[index_list]['start']\n",
    "    end_list = df.loc[index_list]['end']\n",
    "    \n",
    "    t_list = sorted(list(zip(start_list, end_list)))\n",
    "    print(t_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 20) Determinare, per ogni gene e ogni trascritto il numero di esoni che lo compongono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene     transcript        \n",
       "ARHGAP4  U52112.4-001          23\n",
       "         U52112.4-002          19\n",
       "         U52112.4-003          22\n",
       "         U52112.4-004           7\n",
       "         U52112.4-005           5\n",
       "         U52112.4-006           3\n",
       "         U52112.4-007           2\n",
       "         U52112.4-008           6\n",
       "         U52112.4-009           6\n",
       "         U52112.4-010           7\n",
       "         U52112.4-011          21\n",
       "         U52112.4-012          18\n",
       "         U52112.4-013          17\n",
       "         U52112.4-014          22\n",
       "         U52112.4-015           7\n",
       "         U52112.4-016          10\n",
       "         U52112.4-017           5\n",
       "         U52112.4-018           5\n",
       "         U52112.4-019           5\n",
       "         U52112.4-020           5\n",
       "         U52112.4-021           3\n",
       "         U52112.4-022          14\n",
       "         U52112.4-023           2\n",
       "         U52112.4-024          19\n",
       "ATP6AP1  XX-FW83563B9.4-001     7\n",
       "         XX-FW83563B9.4-002     7\n",
       "         XX-FW83563B9.4-003     5\n",
       "         XX-FW83563B9.4-004     6\n",
       "         XX-FW83563B9.4-006     5\n",
       "AVPR2    U52112.2-001           4\n",
       "         U52112.2-002           4\n",
       "         U52112.2-003           2\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.feature == 'exon'].groupby(['gene', 'transcript'])['feature'].count()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
