{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Manipolazione di un file GTF (Gene Transfer Format) attraverso la libreria `Pandas`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1) Importare `Pandas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 2) Leggere il file GTF\n",
    "\n",
    "    df = pd.read_csv(gtf_file_name, sep='\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('./input.gtf', sep='\\t', header = None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0           1     2      3      4  5  6  7  \\\n",
       "0    ENm006  VEGA_Known  exon  71783  71788  .  -  .   \n",
       "1    ENm006  VEGA_Known   CDS  71783  71788  .  -  0   \n",
       "2    ENm006  VEGA_Known  exon  70312  70440  .  -  .   \n",
       "3    ENm006  VEGA_Known   CDS  70312  70440  .  -  0   \n",
       "4    ENm006  VEGA_Known  exon  69989  70210  .  -  .   \n",
       "..      ...         ...   ...    ...    ... .. .. ..   \n",
       "405  ENm006  VEGA_Known   CDS  56303  56327  .  +  0   \n",
       "406  ENm006  VEGA_Known  exon  56689  57573  .  +  .   \n",
       "407  ENm006  VEGA_Known   CDS  56689  57573  .  +  2   \n",
       "408  ENm006  VEGA_Known  exon  57680  58323  .  +  .   \n",
       "409  ENm006  VEGA_Known   CDS  57680  57885  .  +  2   \n",
       "\n",
       "                                                     8  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**NB**: `read_csv()` ha un attributo `names` che permette di specificare la lista dei nomi delle colonne del data frame restituito."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 3) Cambiare i nomi delle colonne\n",
    "\n",
    "I nomi delle colonne devono essere:\n",
    "- reference\n",
    "- source\n",
    "- feature\n",
    "- start\n",
    "- end\n",
    "- score\n",
    "- strand\n",
    "- frame\n",
    "- attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "replace_dict = {0 : 'reference', 1 : 'source', 2 : 'feature', 3 : 'start', 4 : 'end', 5 : 'score', 6 : 'strand', 7 : 'frame', 8 : 'attributes'}\n",
    "df.rename(columns = replace_dict, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENm006</td>\n",
       "      <td>VEGA_Known</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    reference      source feature  start    end score strand frame  \\\n",
       "0      ENm006  VEGA_Known    exon  71783  71788     .      -     .   \n",
       "1      ENm006  VEGA_Known     CDS  71783  71788     .      -     0   \n",
       "2      ENm006  VEGA_Known    exon  70312  70440     .      -     .   \n",
       "3      ENm006  VEGA_Known     CDS  70312  70440     .      -     0   \n",
       "4      ENm006  VEGA_Known    exon  69989  70210     .      -     .   \n",
       "..        ...         ...     ...    ...    ...   ...    ...   ...   \n",
       "405    ENm006  VEGA_Known     CDS  56303  56327     .      +     0   \n",
       "406    ENm006  VEGA_Known    exon  56689  57573     .      +     .   \n",
       "407    ENm006  VEGA_Known     CDS  56689  57573     .      +     2   \n",
       "408    ENm006  VEGA_Known    exon  57680  58323     .      +     .   \n",
       "409    ENm006  VEGA_Known     CDS  57680  57885     .      +     2   \n",
       "\n",
       "                                            attributes  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4) Eliminare le colonne `source` e `score` e sostituire l'identificatore `ENm006` con l'identificatore `ENCODE_REGION` in tutti i campi della colonna `reference`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop(['source', 'score'], axis=1, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['reference'] = 'ENCODE_REGION'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end strand frame  \\\n",
       "0    ENCODE_REGION    exon  71783  71788      -     .   \n",
       "1    ENCODE_REGION     CDS  71783  71788      -     0   \n",
       "2    ENCODE_REGION    exon  70312  70440      -     .   \n",
       "3    ENCODE_REGION     CDS  70312  70440      -     0   \n",
       "4    ENCODE_REGION    exon  69989  70210      -     .   \n",
       "..             ...     ...    ...    ...    ...   ...   \n",
       "405  ENCODE_REGION     CDS  56303  56327      +     0   \n",
       "406  ENCODE_REGION    exon  56689  57573      +     .   \n",
       "407  ENCODE_REGION     CDS  56689  57573      +     2   \n",
       "408  ENCODE_REGION    exon  57680  58323      +     .   \n",
       "409  ENCODE_REGION     CDS  57680  57885      +     2   \n",
       "\n",
       "                                            attributes  \n",
       "0     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "1     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "2     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "3     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "4     transcript_id \"U52112.4-005\"; gene_id \"ARHGAP4\";  \n",
       "..                                                 ...  \n",
       "405  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "406  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "407  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "408  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "409  transcript_id \"U52112.2-001\"; gene_id \"AVPR2\";...  \n",
       "\n",
       "[410 rows x 7 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 5) Sostituire la colonna degli attributi con le due colonne  `transcript` e `gene`\n",
    "\n",
    "La colonne `transcript` e `gene` dovranno contenere solo l'ID del trascritto e del gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['gene'] = ''\n",
    "df['transcript'] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "for (index, record) in df.iterrows():\n",
    "    transcript_id = re.search('transcript_id\\s+(.+?);', record['attributes']).group(1).replace('\"', '')\n",
    "    gene_id = re.search('gene_id\\s+(.+?);', record['attributes']).group(1).replace('\"', '')\n",
    "    df.loc[index, 'transcript'] =  transcript_id\n",
    "    df.loc[index, 'gene'] =  gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.drop('attributes', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end strand frame     gene    transcript\n",
       "0    ENCODE_REGION    exon  71783  71788      -     .  ARHGAP4  U52112.4-005\n",
       "1    ENCODE_REGION     CDS  71783  71788      -     0  ARHGAP4  U52112.4-005\n",
       "2    ENCODE_REGION    exon  70312  70440      -     .  ARHGAP4  U52112.4-005\n",
       "3    ENCODE_REGION     CDS  70312  70440      -     0  ARHGAP4  U52112.4-005\n",
       "4    ENCODE_REGION    exon  69989  70210      -     .  ARHGAP4  U52112.4-005\n",
       "..             ...     ...    ...    ...    ...   ...      ...           ...\n",
       "405  ENCODE_REGION     CDS  56303  56327      +     0    AVPR2  U52112.2-001\n",
       "406  ENCODE_REGION    exon  56689  57573      +     .    AVPR2  U52112.2-001\n",
       "407  ENCODE_REGION     CDS  56689  57573      +     2    AVPR2  U52112.2-001\n",
       "408  ENCODE_REGION    exon  57680  58323      +     .    AVPR2  U52112.2-001\n",
       "409  ENCODE_REGION     CDS  57680  57885      +     2    AVPR2  U52112.2-001\n",
       "\n",
       "[410 rows x 8 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 6) Aggiungere la colonna `length` contenente la lunghezza della feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['length'] = df['end'] - df['start'] + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reindicizzazione delle colonne:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.reindex(columns = ['reference', 'feature', 'start', 'end', 'length', 'strand', 'frame', 'gene', 'transcript'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>71783</td>\n",
       "      <td>71788</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>70312</td>\n",
       "      <td>70440</td>\n",
       "      <td>129</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>69989</td>\n",
       "      <td>70210</td>\n",
       "      <td>222</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ARHGAP4</td>\n",
       "      <td>U52112.4-005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>405</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56303</td>\n",
       "      <td>56327</td>\n",
       "      <td>25</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>406</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>407</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>56689</td>\n",
       "      <td>57573</td>\n",
       "      <td>885</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>exon</td>\n",
       "      <td>57680</td>\n",
       "      <td>58323</td>\n",
       "      <td>644</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ENCODE_REGION</td>\n",
       "      <td>CDS</td>\n",
       "      <td>57680</td>\n",
       "      <td>57885</td>\n",
       "      <td>206</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>AVPR2</td>\n",
       "      <td>U52112.2-001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>410 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reference feature  start    end  length strand frame     gene  \\\n",
       "0    ENCODE_REGION    exon  71783  71788       6      -     .  ARHGAP4   \n",
       "1    ENCODE_REGION     CDS  71783  71788       6      -     0  ARHGAP4   \n",
       "2    ENCODE_REGION    exon  70312  70440     129      -     .  ARHGAP4   \n",
       "3    ENCODE_REGION     CDS  70312  70440     129      -     0  ARHGAP4   \n",
       "4    ENCODE_REGION    exon  69989  70210     222      -     .  ARHGAP4   \n",
       "..             ...     ...    ...    ...     ...    ...   ...      ...   \n",
       "405  ENCODE_REGION     CDS  56303  56327      25      +     0    AVPR2   \n",
       "406  ENCODE_REGION    exon  56689  57573     885      +     .    AVPR2   \n",
       "407  ENCODE_REGION     CDS  56689  57573     885      +     2    AVPR2   \n",
       "408  ENCODE_REGION    exon  57680  58323     644      +     .    AVPR2   \n",
       "409  ENCODE_REGION     CDS  57680  57885     206      +     2    AVPR2   \n",
       "\n",
       "       transcript  \n",
       "0    U52112.4-005  \n",
       "1    U52112.4-005  \n",
       "2    U52112.4-005  \n",
       "3    U52112.4-005  \n",
       "4    U52112.4-005  \n",
       "..            ...  \n",
       "405  U52112.2-001  \n",
       "406  U52112.2-001  \n",
       "407  U52112.2-001  \n",
       "408  U52112.2-001  \n",
       "409  U52112.2-001  \n",
       "\n",
       "[410 rows x 9 columns]"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 7) Rimuovere dal *data frame* tutte le features di lunghezza minore o uguale a 6 basi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 8) Estrarre il *data frame* dei primi 20 esoni più lunghi e degli ultimi 20 esoni localizzati sul *reference*\n",
    "\n",
    "*SUGGERIMENTO*: usare la funzione:\n",
    "\n",
    "    pd.merge(first_df, second_df, how='outer')\n",
    "    \n",
    "per unire verticalmente due *data frame* con le stesse colonne."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 9) Estrarre la lista dei geni presenti nel file GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 10) Estrarre la lista dei trascritti presenti nel file GTF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 11) Determinare, per ogni gene, la lunghezza media, massima e minima degli esoni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 12) Determinare la lunghezza minima degli esoni ed estrarre tutti i trascritti che contengono un esone di lunghezza minima"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 13) Contare quanti trascritti sono annotati per il gene `ARHGAP4`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 14) Estrarre la lista dei geni con strand `+`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 15) Estrarre gli esoni (distinti) del gene `ATP6AP1` in una lista di tuple (start, end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 16) Contare il numero di trascritti del gene `ARHGAP4` che hanno una CDS annotata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 17) Estrarre lo strand del gene  `ATP6AP1`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 18) Determinare il trascritto che ha più esoni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 19) Estrarre per ogni trascritto del gene  `ARHGAP4` la lista delle tuple (start, end) dei suoi esoni ordinate per start crescente"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 20) Determinare, per ogni gene e ogni trascritto il numero di esoni che lo compongono"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
